crates/pico_structures/src/world.rs

```diff
@@ -57,21 +68,236 @@ impl World {
+        let mut block_light_volume: Vec<u8> = vec![0u8; volume_size];
+
+        let get_index = |x: i32, y: i32, z: i32| -> usize {
+            (y * world_width * world_length + z * world_width + x) as usize
+        };
+
+        for x in 0..world_width {
+            for z in 0..world_length {
+                let mut sky_light = 15u8;
+                for y in (0..total_height).rev() {
+                    let pos = Coordinates::new(x, y, z);
+                    let idx = get_index(x, y, z);
+                    if schematic.is_transparent(pos) {
+                        sky_light_volume[idx] = sky_light;
+                    } else {
+                        sky_light_volume[idx] = 0;
+                        sky_light = 0;
+                    }
+                }
+            }
+        }
+
+        for x in 0..world_width {
+            for z in 0..world_length {
+                for y in 0..total_height {
+                    let pos = Coordinates::new(x, y, z);
+                    let emitted = schematic.get_emitted_light(pos);
+                    if emitted > 0 {
+                        let idx = get_index(x, y, z);
+                        block_light_volume[idx] = emitted;
+                    }
+                }
+            }
+        }
+
+        Self::propagate_light(&mut sky_light_volume, schematic, world_width, world_length, total_height, &get_index);
+        Self::propagate_light(&mut block_light_volume, schematic, world_width, world_length, total_height, &get_index);
+
+        let sky_light_by_chunk = Self::convert_to_chunk_sections(
+            &sky_light_volume, size_in_chunks, chunk_column_count, world_width, world_length, total_height, &get_index, 15
+        );
+        let block_light_by_chunk = Self::convert_to_chunk_sections(
+            &block_light_volume, size_in_chunks, chunk_column_count, world_width, world_length, total_height, &get_index, 0
+        );
+
+        (sky_light_by_chunk, block_light_by_chunk)
+    }
+
+    /// Propagate light through transparent blocks using parallel processing
+    fn propagate_light<F>(
+        light_volume: &mut [u8],
+        schematic: &Schematic,
+        world_width: i32,
+        world_length: i32,
+        total_height: i32,
+        get_index: &F,
+    ) where
+        F: Fn(i32, i32, i32) -> usize + Sync,
+    {
+        let total_blocks = (world_width * world_length * total_height) as usize;
+
+        for _ in 0..15 {
+            let new_values: Vec<(usize, u8)> = (0..total_blocks)
+                .into_par_iter()
+                .filter_map(|i| {
+                    let x = (i as i32) % world_width;
+                    let z = ((i as i32) / world_width) % world_length;
+                    let y = (i as i32) / (world_width * world_length);
+
+                    let idx = get_index(x, y, z);
+                    let current = light_volume[idx];
+
+                    let current_pos = Coordinates::new(x, y, z);
```

Comment:
I'm thinking of a constructor `Coordinates::from_index` that converts the index to x,y,z